make/program

```diff
@@ -23,22 +24,26 @@ ifneq ($(findstring allow_undefined,$(STANCFLAGS)),)
 $(STAN_TARGETS) examples/bernoulli/bernoulli$(EXE) $(patsubst %.stan,%$(EXE),$(wildcard src/test/test-models/*.stan)) : CXXFLAGS_PROGRAM += -include $(USER_HEADER)
 endif
 
-
 %.hpp : %.stan bin/stanc$(EXE)
 	@echo ''
 	@echo '--- Translating Stan model to C++ code ---'
 	$(WINE) bin/stanc$(EXE) $(STANCFLAGS) --o=$@ $<
 
+%.d: %.hpp
 
 .PRECIOUS: %.hpp
-%$(EXE) : %.hpp $(CMDSTAN_MAIN) $(LIBSUNDIALS) $(MPI_TARGETS)
+%$(EXE) : %.hpp $(CMDSTAN_MAIN_O) $(LIBSUNDIALS) $(MPI_TARGETS)
 	@echo ''
-	@echo '--- Linking C++ model ---'
-	$(LINK.cpp) $(CXXFLAGS_PROGRAM) -include $< $(CMDSTAN_MAIN) $(LDLIBS) $(LIBSUNDIALS) $(MPI_TARGETS) $(OUTPUT_OPTION)
+	@echo '--- Compiling, linking C++ code ---'
+	cp $*.hpp $*.cpp; \
+    $(LINK.cpp) $(CXXFLAGS_PROGRAM) -c -o $*.o $*.cpp ; \
```

Comment:
Oh.. this is less ideal, I think. You can make the compiler interpret the hpp file just like a c++ file with the `-x c++` option. So I would suggest to replace 38-41 with this:

```
        $(COMPILE.cpp) $(CXXFLAGS_PROGRAM) -x c++ -o $*.o $<
        $(LINK.cpp) $(LDLIBS) $(LIBSUNDIALS) $(MPI_TARGETS) $(CMDSTAN_MAIN_O) $*.o $(OUTPUT_OPTION)
```

This way you pickup the right things for compilation (first line) and for linking (second line). Moreover, you do not need to do this copying and removing business.